### ISSUE TITLE
Verification of Ed25519 certificates

### ISSUE THREAD

**Issue by fspreiss** [rusticata/x509-parser#71]
Hi! We would like to use `X509Certificate::verify_signature` to verify the signature of a certificate created with OpenSSL that was (self-)signed with [an Ed25519 key](https://docs.rs/openssl/0.10.32/openssl/pkey/struct.PKey.html#method.generate_ed25519).

Currently this fails with an `X509Error::SignatureUnsupportedAlgorithm` because the respective OID (1.3.101.112; see also [RFC 8410](https://tools.ietf.org/html/rfc8410)) is [neither supported by the `x509-parser`](https://docs.rs/x509-parser/0.9.1/src/x509_parser/certificate.rs.html#161-176) nor [by the `oid-registry`](https://github.com/rusticata/oid-registry/blob/master/assets/oid_db.txt).

It seems the underlying `ring` library, however, does support Ed25519 via `ring::signature::ED25519`, which also implements the required `ring::signature::VerificationAlgorithm`, so it seems this is a matter of extending the `oid-registry` and wiring all the existing pieces together.

Would you be interested in supporting (the more and more popular) Ed25519, and if yes, what would be a sensible path forward? For example, is the best/fastest way to submit respective PRs to `oid-registry` and then to `x509-parser`? Thanks!

**Comment by chifflier** [comment]
Hi,
Ed25519 is definitely something we'd like to support, and that I'd be happy to add.
The only open question at this point is whether to switch the verification layer to openssl instead of ring (See #72 ). Since both support it, I'll try to add support of Ed25519 as soon as possible, but it may have to wait for the implementation choice.
Thanks!

**Comment by fspreiss** [comment]
@chifflier, happy to hear that Ed25519 is something you'd like to support.

The compatibility argument for openssl that you make in #72 is certainly a valid one.

For projects like ours, having openssl as dependency would not work, however, because we compile our crate to WebAssembly (with `--target wasm32-unknown-unknown`).

We found a good workaround for verifying our Ed25519 certificates by verifying the signature ourselves with the [ed25519_dalek](https://crates.io/crates/ed25519-dalek) crate: the [x509_parser::certificate::X509Certificate](https://docs.rs/x509-parser/0.9.1/x509_parser/certificate/struct.X509Certificate.html)'s `signature_value` and `tbs_certificate` contained everything we needed, so thanks for that.

**Comment by ThibsG** [comment]
@fspreiss, were you able to check certificate by using [`verify_signature`](https://docs.rs/x509-parser/0.9.2/x509_parser/certificate/struct.X509Certificate.html#method.verify_signature) from `x509-parser` crate in WASM?
If yes, didn't you have some troubles with incomplete WASM `ring` support?

**Comment by fspreiss** [comment]
@ThibsG, because back in March when we needed this feature, `X509Certificate::verify_signature` didn't support Ed25519 yet, so we opted for a workaround by verifying the signature ourselves with the [ed25519_dalek](https://crates.io/crates/ed25519-dalek) crate (the [x509_parser::certificate::X509Certificate](https://docs.rs/x509-parser/0.9.1/x509_parser/certificate/struct.X509Certificate.html)'s `signature_value` and `tbs_certificate` contained everything we needed).

We therefore never got to a point where we used `X509Certificate::verify_signature` (and therefore indirectly `ring`) in WASM, so I cannot speak to whether this works or not. I can, however, say for sure that the verification with `ed25519_dalek` works in WASM.
